### ISSUE TITLE
Support Dostoevsky compaction scheme

### ISSUE THREAD

**Issue by criccomini** [slatedb/slatedb#168]
We should add the [dostoevsky](https://stratos.seas.harvard.edu/sites/scholar.harvard.edu/files/stratos/files/dostoevskykv.pdf) compaction improvement where you always compact into SR 0 if you are doing a compaction that produces a run that is similarly sized to SR 0. (SR 0 being the oldest/base sorted range).

/cc @rodesai for additional details

**Comment by criccomini** [comment]
@rodesai Correct me if I'm wrong, but I believe you were talking about this part of the paper, right?

> Lazy Leveling is a merge policy that eliminates merging at all but the largest level of LSM-tree.

**Comment by rodesai** [comment]
@criccomini yes that's right. In our DB the equivalent thing to do would be to do the following for any compaction that the scheduler picks:
1. check if the the compaction is adjacent to SR 0
2. estimate the resulting size of the compaction (currently the estimate is just the sum of the sources' sizes). If the resulting size produces a sorted run that can be compacted with SR 0, then include SR 0 in the compaction

**Comment by rodesai** [comment]
Couple more points:
- the high-level idea here is that you're trading off some added write amplification from the more aggressive compaction of the largest sorted run for much less space amplification. For more details see https://scholar.harvard.edu/files/stratos/files/dostoevskykv.pdf

Second, this is useful for workloads where there are expected to be lots of updates. For insert-dominant workloads, this doesn't always make sense to do. The "always" caveat comes from the perspective that for insert-dominant workloads you probably need some sort of expiry mechanism, otherwise your db will always grow. You could implement expiry in a few ways:
- by writing tombstones for keys when they're expired. In this case the tombstones are essentially updates and the db would benefit from the lazy-leveling optimization.
- by rolling the db at some interval - e.g. you initialize a new "db" at some fixed interval, and then delete the whole db when it's contents are expired. This could be done externally by the client, or we could opt to do this in slatedb (akin to time-windowed compaction in scylladb (https://opensource.docs.scylladb.com/stable/architecture/compaction/compaction-strategies.html#time-window-compaction-strategy-twcs) . In this case, there are no updates and it doesn't make sense to enable this optimization.
But the tl;dr for this bullet is that this behaviour should be optional.
